amcl/src/amcl_node.cpp

```diff
@@ -1587,11 +1591,28 @@ AmclNode::laserReceived(const sensor_msgs::LaserScanConstPtr& laser_scan)
 
         pure_amcl_delta = getAmclMovement(p, base_time, reliable_pose_msg);
         pure_odom_delta = getOdomMovement(pose, p.header.stamp, base_time);
-        double dx = sqrt(std::pow(pure_odom_delta.v[0] - pure_amcl_delta.v[0], 2) +
-                         std::pow(pure_odom_delta.v[1] - pure_amcl_delta.v[1], 2));
-        double dz = angle_diff(pure_odom_delta.v[2], pure_amcl_delta.v[2]);
-        bool acceptable_x = std::fabs(dx) < odom_amcl_x_thre_;
-        bool acceptable_z = std::fabs(dz) < odom_amcl_z_thre_;
+        
+        // When the tf of map-odom is not obtained in time, the tf of map-baselink is not obtained as well,
+        // so pure_odom_delta is 0. In that case, pure_amcl_delta is substituted for pure_odom_delta
+        // as a bitter measure.
+        if (pure_odom_delta.v[0] == 0.0 && pure_odom_delta.v[1] == 0.0 && pure_odom_delta.v[2] == 0.0)
+          pure_odom_delta = pure_amcl_delta;
```

Comment:
Need to increase odom cycle?
If we substitute amcl, it is good when it is normal, but when odom or amcl has a strange value, we cannot save what could be saved by correction, and it would be a waste of time.

----

_odomの周期を上げる必要がある?
amclで代用すると、正常時は良いが、odomかamclが変な値になったときに、補正で救えたものを救えなくなり、もったいないのでは。_

Reply:
> 補正で救えたものを救えなくなり、もったいないのでは。
そのとおりですね。

map-odom-base_link の map-odom が間に合わないため、laser_scan の timestamp において、map-base_link が得られないことがある。
ekfはodom, imu, amclを食ってmap-odomを吐くが、
odomの周期を上げることでmap-odomの周期があがるのか。またそれ以外の方法はないか。
amclを食うまでmap-odomを吐かないという場合がやっかい。
map-baselinkではなく、odom-base_linkとmap-amcl_base_linkを比較すれば問題は解決しそうだが、後者はframe同士が違うので、座標変換をどうにか工夫しなければいけない。

という感じです。@yakato-jun さんなにかアイデアあったりしますでしょうか？

Reply:
delay が max 0.03s で、平均してだいたい0.01sぐらいだったので、苦肉の策としてros::Time(0)を使ってもとりあえず問題ないだろうと判断しました。あとで実機でテストします。
元の実装の致命的な点はodom correction をかけるときにdelayが発生してmap-base_link間のtfがとってこれないことにあります。tfがとってこれない例外処理として、移動量を0にしています。なぜか現在再生しているros bag ではこのゼロ補正が頻発しているようなので対策が必要となりました。ros::Time(0)だとそれがかなり緩和されました。

もし、なにかほかによいアイデアがあれば教えていただきたいです。